### PR TITLE
feat: TCP Client Server Half Duplex Communication

### PR DESCRIPTION
**Pull request by NVombat**
#### Description of Change

Added the server and client code for half duplex TCP communication. Ran the code to make sure it compiles and runs successfully. Commented the code and followed contributing guidelines.

#### References

Implements Half Duplex Communication for #852 

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C/pull/853"></a>

### CODE REVIEW

**Comment by NVombat** [comment]
Is the last commit i made fine ?? I documented each part of the code carefully so every line and function is explained

File: client_server/tcp_half_duplex_client.c [hunk 1]
```diff
@@ -0,0 +1,145 @@
+/**
+ * @file
+ * @author [NVombat](https://github.com/NVombat)
+ * @brief Client-side implementation of [TCP Half Duplex
+ * Communication](http://www.tcpipguide.com/free/t_SimplexFullDuplexandHalfDuplexOperation.htm)
+ * @see tcp_half_duplex_client.c
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_client.c:6]:
```suggestion
 * @see tcp_half_duplex_server.c
```

File: client_server/tcp_half_duplex_client.c [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+It prints str and an implementation-defined error
+message corresponding to the global variable errno.
+*/
+void error()
+{
+    perror("Socket Creation Failed");
+    exit(EXIT_FAILURE);
+}
+
+/* MAIN DRIVER CODE */
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_client.c:41]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

File: client_server/tcp_half_duplex_client.c [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+#include <sys/types.h>  /// For definitions to allow for the porting of BSD programs
+#include <unistd.h>  /// For miscellaneous symbolic constants and types, and miscellaneous functions
+
+#define PORT 8100  /// Define port over which communication will take place
+
+/*
+Function is used to print an error message to stderr.
+It prints str and an implementation-defined error
+message corresponding to the global variable errno.
+*/
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_client.c:34]:
```suggestion
/**
 * @brief Utility function used to print an error message to `stderr`.
 * It prints `str` and an implementation-defined error
 * message corresponding to the global variable `errno`.
 * @returns void
 */
```

File: client_server/tcp_half_duplex_server.c [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+ * The algorithm is based on the simple TCP client and server model. However,
+ * instead of the server only sending and the client only receiving data,
+ * The server and client can both send data but only one at a time. This is
+ * implemented by using a particular ordering of the `send()` and `recv()`
+ * functions. When one of the clients or servers is sending, the other can only
+ * receive and vice-versa. In this way, the Half Duplex Form of communication
+ * can be represented using the TCP server-client model & socket programming
+ */
+
+// Imports
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_server.c:18]:
```suggestion
```

File: client_server/tcp_half_duplex_server.c [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+#include <sys/types.h>  /// For definitions to allow for the porting of BSD programs
+#include <unistd.h>  /// For miscellaneous symbolic constants and types, and miscellaneous functions
+
+#define PORT 8100  /// Define port over which communication will take place
+
+/*
+Function is used to print an error message to stderr.
+It prints str and an implementation-defined error
+message corresponding to the global variable errno.
+*/
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_server.c:34]:
```suggestion
/**
 * @brief Utility function used to print an error message to `stderr`.
 * It prints `str` and an implementation-defined error
 * message corresponding to the global variable `errno`.
 * @returns void
 */
```

File: client_server/tcp_half_duplex_server.c [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+void error()
+{
+    perror("Socket Creation Failed");
+    exit(EXIT_FAILURE);
+}
+
+/* MAIN DRIVER CODE */
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_server.c:41]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: client_server/tcp_half_duplex_client.c [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+ * The algorithm is based on the simple TCP client and server model. However,
+ * instead of the server only sending and the client only receiving data,
+ * the server and client can both send data but only one at a time. This is
+ * implemented by using a particular ordering of the `send()` and `recv()`
+ * functions. When one of the clients or servers is sending, the other can only
+ * receive and vice-versa. In this way, the Half Duplex Form of communication
+ * can be represented using the TCP server-client model & socket programming
+ */
+
+// Imports
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_client.c:18]:
```suggestion
```

File: client_server/tcp_half_duplex_server.c (hunk 4 above)
Comment (Panquesito7) [inline client_server/tcp_half_duplex_server.c:18]:
```suggestion
```

File: client_server/tcp_half_duplex_client.c [hunk 8]
```diff
@@ -0,0 +1,149 @@
+/**
+ * @file
+ * @author [NVombat](https://github.com/NVombat)
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_client.c:3]:
```suggestion
 * @author [Nikhill Vombatkere](https://github.com/NVombat)
```

File: client_server/tcp_half_duplex_server.c [hunk 9]
```diff
@@ -0,0 +1,171 @@
+/**
+ * @file
+ * @author [NVombat](https://github.com/NVombat)
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_server.c:3]:
```suggestion
 * @author [Nikhill Vombatkere](https://github.com/NVombat)
```

**Comment by NVombat** [comment]
Are there any other changes that need to be made ?

File: client_server/tcp_half_duplex_client.c [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+{
+    /** Variable Declarations */
+    int sockfd;  // Socket descriptors - Like file handles but for sockets
+    struct sockaddr_in
+        server_addr;  // Basic structures for all syscalls and functions that
+                      // deal with internet addresses. Structures for handling
+                      // internet addresses
+    char serverResponse[10000],
+        clientResponse[10000];  // Character arrays to read and store string
+                                // data for communication
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_client.c:55]:
```suggestion
    int sockfd;  ///< socket descriptors - Like file handles but for sockets
    struct sockaddr_in
        server_addr;  ///< basic structures for all syscalls and functions that
                      /// deal with internet addresses. Structures for handling
                      /// internet addresses
    char serverResponse[10000],
        clientResponse[10000];  ///< Character arrays to read and store string
                                ///< data for communication
```

File: client_server/tcp_half_duplex_server.c [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+{
+    /** Variable Declarations */
+    int sockfd, conn;  // Socket descriptors - Like file handles but for sockets
+    char server_msg[10000],
+        client_msg[10000];  // Character arrays to read and store string data
+                            // for communication
+    struct sockaddr_in server_addr,
+        client_addr;  // Basic structures for all syscalls and functions that
+                      // deal with internet addresses. Structures for handling
+                      // internet addresses
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_server.c:55]:
```suggestion
    int sockfd, conn;  ///< socket descriptors - Like file handles but for sockets
    char server_msg[10000],
        client_msg[10000];  ///< character arrays to read and store string data
                            /// for communication
    struct sockaddr_in server_addr,
        client_addr;  ///< asic structures for all syscalls and functions that
                      /// deal with internet addresses. Structures for handling
                      /// internet addresses
```

**Comment by NVombat** [comment]
I updated all comment formatting to be uniform.

File: client_server/tcp_half_duplex_client.c [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+     * without record boundaries. Aside from the bidirectionality of data flow,
+     * a pair of connected stream sockets provides an interface nearly identical
+     * to pipes.
+     *
+     * 0 (Protocol) - Specifies a particular protocol to be used with the
+     * socket. Specifying a protocol of 0 causes socket() to use an unspecified
+     * default protocol appropriate for the requested socket type.
+     */
+    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0)
+        error();
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_client.c:74]:
```suggestion
    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
        error();
    }
```

File: client_server/tcp_half_duplex_server.c [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+     * a pair of connected stream sockets provides an interface nearly identical
+     * to pipes
+     *
+     * 0 (Protocol) - Specifies a particular protocol to be used with the
+     * socket. Specifying a protocol of 0 causes socket() to use an unspecified
+     * default protocol appropriate for the requested socket type
+     */
+    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0)
+        error();  ///< Error if the socket descriptor has a value lower than 0 -
+                  /// socket wasnt created
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_server.c:76]:
```suggestion
    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
        error();  ///< Error if the socket descriptor has a value lower than 0 -
                  /// socket wasnt created
    }
```

File: client_server/tcp_half_duplex_server.c [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+    server_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+
+    printf("Server is running...\n");
+
+    /**
+     * This binds the socket descriptor to the server thus enabling the server
+     * to listen for connections and communicate with other clients
+     */
+    if (bind(sockfd, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0)
+        error();  /// If binding is unsuccessful
```
Comment (Panquesito7) [inline client_server/tcp_half_duplex_server.c:110]:
```suggestion
    if (bind(sockfd, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
        error();  /// If binding is unsuccessful
    }
```

File: client_server/tcp_half_duplex_client.c [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main()
+{
+    /** Variable Declarations */
+    int sockfd;  ///< socket descriptors - Like file handles but for sockets
```
[the remark this answers is not in the document]
Comment (NVombat) [reply client_server/tcp_half_duplex_client.c:48]: Okay i will look into this and make the change.

**Review by Panquesito7 — APPROVED** [review]
Amazing work, @NVombat! It's been great to be your first contribution here! We hope you keep contributing! 😄👍🎉

**Comment by NVombat** [comment]
Thank you! Will contribute more for sure!

**Comment by NVombat** [comment]
Just wanted to know - By when will this get merged??

**Comment by Panquesito7** [comment]
@amino19, would you like to review this PR? Thanks. 🙂

@ayaankhan98 and @mishraabhinn, please wait until @amino19 reviews this PR.
They're practicing with us and learning to review PRs with us. 🙂

**Review by ayaankhan98 — APPROVED** [review]
LGTM!

**Review by amino19 — APPROVED** [review]
@NVombat: Thanks for your contributions in **The Algorithms**!
LGTM! :partying_face: 

@Panquesito7: Pl take this to your end! : )

**Pull request merged**